internal/pkg/gateway/registry.go

```diff
@@ -264,63 +266,54 @@ func (reg *registry) orderers(channel string) ([]*orderer, error) {
 	if err != nil {
 		return nil, err
 	}
-	var orderers []*orderer
 
-	// Get the config
-	config, err := reg.discovery.Config(channel)
-	if err != nil {
-		return nil, err
-	}
-
-	reg.configLock.RLock()
-	defer reg.configLock.RUnlock()
-
-	for _, eps := range config.GetOrderers() {
-		for _, ep := range eps.Endpoint {
-			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
-			if ord, ok := reg.broadcastClients[url]; ok {
-				orderers = append(orderers, ord)
-			}
-		}
+	var orderers []*orderer
+	for _, url := range reg.channelOrderers[channel] {
+		orderers = append(orderers, reg.broadcastClients[url])
 	}
 
 	return orderers, nil
 }
 
 func (reg *registry) registerChannel(channel string) error {
-	// todo need to handle membership updates
 	reg.configLock.Lock() // take a write lock to populate the registry maps
 	defer reg.configLock.Unlock()
 
-	if reg.channelsInitialized[channel] {
-		return nil
-	}
-	// get config and peer discovery info for this channel
-
-	// Get the config
-	config, err := reg.discovery.Config(channel)
-	if err != nil {
-		return err
-	}
-	// get the tlscerts
-	for msp, info := range config.GetMsps() {
-		reg.tlsRootCerts[msp] = info.GetTlsRootCerts()
-	}
+	if valid, exists := reg.configValid.Load(channel); !exists || !valid.(bool) {
+		// Get the config
+		config, err := reg.discovery.Config(channel)
+		if err != nil {
+			return err
+		}
+		// get the tlscerts
+		for msp, info := range config.GetMsps() {
+			reg.tlsRootCerts[msp] = info.GetTlsRootCerts()
+		}
 
-	for mspid, eps := range config.GetOrderers() {
-		for _, ep := range eps.Endpoint {
-			address := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
-			if _, ok := reg.broadcastClients[address]; !ok {
-				// this orderer is new - connect to it and add to the broadcastClients registry
-				tlsRootCerts := reg.tlsRootCerts[mspid]
-				orderer, err := reg.endpointFactory.newOrderer(address, mspid, tlsRootCerts)
-				if err != nil {
-					return err
+		// Get the current set of orderers.  Use existing connections, if available.  Some may have been added, some removed.
+		var channelOrderers []string
+		for mspid, eps := range config.GetOrderers() {
+			for _, ep := range eps.Endpoint {
+				address := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
+				if _, ok := reg.broadcastClients[address]; !ok {
+					// this orderer is new - connect to it and add to the broadcastClients registry
+					tlsRootCerts := reg.tlsRootCerts[mspid]
+					newOrd, err := reg.endpointFactory.newOrderer(address, mspid, tlsRootCerts)
```

Comment:
I know this code was there previously and isn't introduced in this commit. 
Regardless, you may want to know that there is a problem with the current code, and you may want to fix it.

the **newOrderer** method will be called upon every channel config, and that method calls **newConnection** which in turn performs a **synchronous** gRPC **Dial**.
This means that if there is an orderer which you have never connected to before, the method **Submit** of the gateway will **freeze** for the duration of the timeout. 

Perhaps it would be better to make the gRPC connection asynchronous and let gRPC handle it, via the **AsyncConnect** flag.

Another problem is that I don't see anywhere that you close connections to orderers that do not exist anymore. 
It can be easily inferred from the fact that the gRPC connection pointer is swallowed by a gRPC client, and can never be accessed ever again.

Reply:
By the way, there is an even worse problem with the endorser approach, and if a peer is simply shut down forever, you will have a file descriptor leak because the gRPC connection can never be closed from the code.

Reply:
For now, I've kept the connection synchronous, and I'll create a separate issue to change it to asynchronous.  That will require some extra work (not related to this commit) in order to potentially wait on connections when they are needed.

I'll also create an issue to handle connection closure.  Again, beyond the scope of this commit.